File: platforms/android/java/app/msime/client/InputDiagnosticPolicy.java
package app.msime.client;

/** Bounds transient Engine diagnostics before they reach the keyboard surface. */
public final class InputDiagnosticPolicy {
    public static final long DISMISS_DELAY_MILLIS = 4_000L;
    public static final int MAX_LENGTH = 1_024;

    private InputDiagnosticPolicy() {}

    public static String normalize(String value) {
        if (value == null) return "";
        String normalized = value.trim();
        if (normalized.isEmpty()) return "";
        if (normalized.length() <= MAX_LENGTH) return normalized;
        return normalized.substring(0, MAX_LENGTH - 1) + "…";
    }

    public static boolean visible(String value) {
        return !normalize(value).isEmpty();
    }
}

File: platforms/harmony/entry/src/main/ets/keyboard/input/InputDiagnosticPolicy.ts
/**
 * Bounds transient Engine diagnostics before they reach the keyboard surface, ported from
 * platforms/android/java/app/msime/client/InputDiagnosticPolicy.java.
 */
export class InputDiagnosticPolicy {
  static readonly DISMISS_DELAY_MILLIS: number = 4000;
  static readonly MAX_LENGTH: number = 1024;

  static normalize(value: string | null): string {
    if (value === null) {
      return '';
    }
    const normalized: string = value.trim();
    if (normalized.length === 0) {
      return '';
    }
    if (normalized.length <= InputDiagnosticPolicy.MAX_LENGTH) {
      return normalized;
    }
    return normalized.substring(0, InputDiagnosticPolicy.MAX_LENGTH - 1) + '…';
  }

  static visible(value: string | null): boolean {
    return InputDiagnosticPolicy.normalize(value).length > 0;
  }
}
